beta/src/content/apis/react-dom/components/input.md

```diff
@@ -0,0 +1,590 @@
+- An input can't be both controlled and uncontrolled at the same time.
+- An input cannot switch between being controlled or uncontrolled over its lifetime.
+- Every controlled input needs an `onChange` event handler that synchronously updates its backing value.
+
+---
+
+## Troubleshooting {/*troubleshooting*/}
+
+### My text input doesn't update when I type into it {/*my-text-input-doesnt-update-when-i-type-into-it*/}
+
+If you render an input with `value` but no `onChange`, you will see an error in the console:
+
+```js
+// 🔴 Bug: controlled text input with no onChange handler
+<input value={something} />
+```
+
+<ConsoleBlock level="error">
+
+You provided a `value` prop to a form field without an `onChange` handler. This will render a read-only field. If the field should be mutable use `defaultValue`. Otherwise, set either `onChange` or `readOnly`.
+
+</ConsoleBlock>
+
+As the error message suggests, if you only wanted to [specify the *initial* value,](#providing-an-initial-value-for-an-input) pass `defaultValue` instead:
+
+```js
+// ✅ Good: uncontrolled input with an initial value
+<input defaultValue={something} />
+```
+
+If you want [to control this input with a state variable,](#controlling-an-input-with-a-state-variable) specify an `onChange` handler:
+
+```js
+// ✅ Good: controlled input with onChange
+<input value={something} onChange={e => setSomething(e.target.value)} />
+```
+
+If the value is intentionally read-only, add a `readOnly` prop to suppress the error:
+
+```js
+// ✅ Good: readonly controlled input without on change
+<input value={something} readOnly={true} />
+```
+
+---
+
+### My checkbox doesn't update when I click on it {/*my-checkbox-doesnt-update-when-i-click-on-it*/}
+
+If you render a checkbox with `checked` but no `onChange`, you will see an error in the console:
+
+```js
+// 🔴 Bug: controlled checkbox with no onChange handler
+<input type="checkbox" checked={something} />
+```
+
+<ConsoleBlock level="error">
+
+You provided a `checked` prop to a form field without an `onChange` handler. This will render a read-only field. If the field should be mutable use `defaultChecked`. Otherwise, set either `onChange` or `readOnly`.
+
+</ConsoleBlock>
+
+As the error message suggests, if you only wanted to [specify the *initial* value,](#providing-an-initial-value-for-an-input) pass `defaultChecked` instead:
+
+```js
+// ✅ Good: uncontrolled checkbox with an initial value
+<input type="checkbox" defaultChecked={something} />
+```
+
+If you want [to control this checkbox with a state variable,](#controlling-an-input-with-a-state-variable) specify an `onChange` handler:
+
+```js
+// ✅ Good: controlled checkbox with onChange
+<input type="checkbox" checked={something} onChange={e => setSomething(e.target.checked)} />
+```
+
+<Pitfall>
+
+You need to read `e.target.checked` rather than `e.target.value` for checkboxes.
+
+</Pitfall>
+
+If the checkbox is intentionally read-only, add a `readOnly` prop to suppress the error:
```

Comment:
FYI [readonly checkboxes/radios aren't really a thing on the web](https://github.com/whatwg/html/issues/2311) and are currently bugged in React: https://github.com/facebook/react/issues/19860